moirae/models/ecm/__init__.py

```diff
@@ -141,28 +162,47 @@ def calculate_terminal_voltage(
         #         + Sum[I_j * R_j(SOC,T)] +
         #         + hyst(SOC,T)
         # Start with OCV
-        Vt = asoh.ocv(soc=transient_state.soc, temp=new_inputs.temperature)
+        Vt = asoh.ocv(soc=transient_state.soc, temp=new_inputs.temperature)  # shape: (ocv_batch, soc_batch, soc_dim=1)
 
         # Add I*R drop ('DCIR')
-        IR_drop = new_inputs.current * asoh.r0.get_value(soc=transient_state.soc,
-                                                         temp=new_inputs.temperature)
-        Vt = Vt + IR_drop
+        IR_drop = new_inputs.current  # shape: (current_batch, 1)
+        r0 = asoh.r0.get_value(soc=transient_state.soc, temp=new_inputs.temperature)  # (r0_batch, soc_batch, soc_dim=1)
+        IR_drop = IR_drop * r0  # broadcasting from right to left gives (r0_batch, trans_batch, soc_dim=1)
+        Vt = Vt + IR_drop  # broadcasting from right to left gives (asoh_batch, trans_batch, soc_dim=1)
 
         # Check series capacitance
         if transient_state.q0 is not None:
-            Vt = Vt + transient_state.q0 / asoh.c0.get_value(soc=transient_state.soc.copy())
+            c0 = asoh.c0.get_value(soc=transient_state.soc.copy())  # shape (c0_batch, soc_batch, soc_dim=1)
+            q0 = transient_state.q0  # shape (q0_batch, 1)
+            deltaV = q0 / c0  # broadcasting from right to left gives (c0_batch, trans_batch, 1)
+            Vt = Vt + deltaV  # broadcasting from right to left gives (asoh_batch, trans_batch, 1)
 
         # Check RC elements
         if transient_state.i_rc.shape[-1] > 0:
             rc_rs = np.array(
                 [rc.r.get_value(soc=transient_state.soc,
                                 temp=new_inputs.temperature)
                  for rc in asoh.rc_elements]
-            )  # Shape: (rc_rs, batch_dim, 1 resistance)
-            V_drops = transient_state.i_rc * rc_rs[:, :, 0].T
-            Vt = Vt + np.sum(V_drops, axis=1, keepdims=True)
+            )  # Shape: (num_rc, rc.r_batch, soc_batch, soc_dim=1)
+            # Transpose to get shape of (rc.r_batch, num_rc, soc_batch, soc_dim=1)
+            rc_rs = np.swapaxes(rc_rs, axis1=0, axis2=1)
+            # Get currents through resistors in RC
+            i_rc = transient_state.i_rc  # shape (i_rc_batch, num_rc)
+            # For proper broadcasting, let's turn this into a shape of (num_rc, i_rc_batch, 1)
+            i_rc = i_rc.T[:, :, None]
+            V_drops = i_rc * rc_rs  # shape (rc.r_batch, num_rc, trans_batch, soc_dim=1)
+            V_drop_sum = np.sum(V_drops, axis=1)  # shape (rc.r_batch, trans_batch, soc_dim=1)
+            Vt = Vt + V_drop_sum  # shape (asoh_batch_size, trans_batch_size, 1)
 
         # Include hysteresis
-        Vt = Vt + transient_state.hyst
+        hyst = transient_state.hyst  # shape (hyst_batch, 1)
```

Comment:
A curiosity point: Are we finding examples where the ASOH and Transient batch sizes are different?

Reply:
Not really. The only case so far was for the computation of the energy of a cell (see new PR)